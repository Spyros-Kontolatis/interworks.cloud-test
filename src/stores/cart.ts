import { defineStore } from "pinia";
import { ref } from "vue";
import type { Cart, CartItem } from "@/types/stores/cart/cart";
import Service from "@/services/Service";

export const useCartStore = defineStore("cart", () => {
  const cart = ref<Cart>({
    id: "",
    items: [],
    totalPrice: 0,
    totalQty: 0,
  });
  const addToCart = (item: CartItem) => {
    const isItemPresent = cart.value.items.find(
      ({ product: { id } }) => id === item.product.id
    );
    if (isItemPresent) {
      isItemPresent.quantity++;
    } else {
      cart.value.items.push(item);
    }
    cart.value.totalQty++;
    cart.value.totalPrice += item.product.price;
    if (cart.value.id) {
      Service.cart().update(cart.value.id, cart.value).execute();
    } else {
      cart.value.id = Date.now().toString();
      Service.cart().create(cart.value).execute();
    }
  };

  const updateItemQuantity = (itemId: string, mode: "remove" | "add") => {
    const itemInCart = cart.value.items.find(
      ({ product: { id } }) => id === itemId
    );
    if (itemInCart) {
      switch (mode) {
        case "add":
          itemInCart.quantity++;
          cart.value.totalQty++;
          cart.value.totalPrice += itemInCart.product.price;
          break;
        case "remove":
          itemInCart.quantity--;
          cart.value.totalQty--;
          cart.value.totalPrice -= itemInCart.product.price;
          break;
      }
    }
    Service.cart().update(cart.value.id, cart.value).execute();
  };

  const removeItemFromCart = (item: CartItem) => {
    cart.value.items = cart.value.items.filter(
      ({ product: { id } }) => id !== item.product.id
    );

    cart.value.totalQty -= item.quantity;
    cart.value.totalPrice -= item.product.price * item.quantity;
    Service.cart().update(cart.value.id, cart.value).execute();
  };
  return {
    cart,
    removeItemFromCart,
    addToCart,
    updateItemQuantity,
  };
});
