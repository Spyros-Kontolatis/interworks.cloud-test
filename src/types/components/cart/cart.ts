import type { CartItem } from "@/types/stores/cart/cart";

export interface CartItemProps {
  item: CartItem;
}
