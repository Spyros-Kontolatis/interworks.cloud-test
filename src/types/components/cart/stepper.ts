export interface StepperProps {
  step: number;
  loading?: boolean;
}
