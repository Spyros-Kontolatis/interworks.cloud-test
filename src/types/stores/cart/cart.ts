import { Product } from "../../components/products/product";

export type CartItem = {
  product: Product;
  quantity: number;
};

export type Cart = {
  items: Array<CartItem>;
  totalQty: number;
  totalPrice: number;
};
