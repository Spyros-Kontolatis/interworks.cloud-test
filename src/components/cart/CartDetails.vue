<template>
  <div class="border border-gray-400 p-4 rounded-md ml-4">
    <div class="flex flex-col">
      <h4 class="font-bold text-start" v-if="cart.items.length">Products:</h4>
      <div
        class="flex justify-between"
        v-for="{ product: { name, id, price }, quantity } in cart.items"
        :key="id"
      >
        <div class="flex grow">
          {{ name }}
          <div class="text-xs flex items-center ml-2">
            ({{ quantity }} x <PriceItem :price="price" />)
          </div>
        </div>
        <div class="text-xs flex gap-2">
          <PriceItem :price="price * quantity" />
        </div>
      </div>
      <div
        class="flex justify-between font-extrabold border-t border-t-gray-400 pt-2"
      >
        <div class="">TOTAL:</div>
        <div>
          <PriceItem :price="cart.totalPrice" />
        </div>
      </div>
    </div>
  </div>
</template>

<script setup lang="ts">
import { useCartStore } from "@/stores/cart";
import { storeToRefs } from "pinia";
import PriceItem from "../product/PriceItem.vue";

const cartStore = useCartStore();
const { cart } = storeToRefs(cartStore);
</script>
