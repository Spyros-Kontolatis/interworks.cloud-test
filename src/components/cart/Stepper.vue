<template>
  <div class="flex">
    <Button
      size="sm"
      color="primary"
      outline
      rounded
      :disabled="step === 1 || loading"
      class="rounded-r-none"
      @click="$emit('update:step', step - 1, 'remove')"
      ><div class="font-bold text-lg">-</div></Button
    >
    <div class="border border-primary py-2 px-4">{{ step }}</div>
    <Button
      size="sm"
      color="primary"
      rounded
      :disabled="loading"
      @click="$emit('update:step', step + 1, 'add')"
      class="rounded-l-none"
      ><div class="font-bold text-lg">+</div></Button
    >
  </div>
</template>

<script setup lang="ts">
import { StepperProps } from "@/types/components/cart/stepper";
import Button from "../base/Button.vue";
import { defineProps, defineEmits } from "vue";

defineProps<StepperProps>();
defineEmits(["update:step"]);
</script>
