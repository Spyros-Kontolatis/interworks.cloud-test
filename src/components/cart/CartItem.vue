<template>
  <div class="border border-gray-400 relative p-4 rounded-md mr-4">
    <div
      v-if="loading"
      class="w-full h-40 absolute top-0 left-0 bg-gray-200 opacity-40"
    >
      <Loader class="w-10 h-10 top-1/3" />
    </div>
    <div class="flex gap-4">
      <div>
        <img :src="item.product.image" width="200" />
      </div>
      <div class="flex grow justify-between text-start">
        <div class="flex flex-col gap-2">
          <div class="font-extrabold">{{ item.product.name }}</div>
          <div>{{ item.product.description }}</div>
          <Stepper
            :loading="loading"
            v-model:step="step"
            @update:step="(_newVal, mode) => updateQuantity(mode)"
          />
        </div>
        <div class="mr-8 flex flex-col">
          <PriceItem :price="item.product.price * item.quantity" />
          <img
            width="20"
            src="../../assets/icons/trash.svg"
            class="cursor-pointer"
            @click="removeItemFromCart(item)"
          />
        </div>
      </div>
    </div>
  </div>
</template>

<script setup lang="ts">
import { defineProps, ref } from "vue";
import type { CartItemProps } from "@/types/components/cart/cart";
import { useCartStore } from "@/stores/cart";
import PriceItem from "../product/PriceItem.vue";
import Loader from "../base/Loader.vue";
import Stepper from "./Stepper.vue";

const props = defineProps<CartItemProps>();
const { updateItemQuantity, removeItemFromCart } = useCartStore();
const step = ref(props.item.quantity);
const loading = ref(false);

const updateQuantity = (mode: "remove" | "add") => {
  loading.value = true;
  updateItemQuantity(props.item.product.id, mode);
  loading.value = false;
};
</script>
