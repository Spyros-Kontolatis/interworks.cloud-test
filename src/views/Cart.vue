<template>
  <div class="flex flex-col md:flex-row h-full">
    <div class="basis-full md:basis-2/3 md:overflow-auto">
      <div v-if="cart.totalQty">
        <h2 class="font-extrabold text-lg text-start">
          {{ cart.totalQty }} Products added to the Cart
        </h2>
        <CartItem
          v-for="item in cart.items"
          :key="`item-${item.product.id}`"
          :item="item"
          class="my-8"
        />
      </div>
      <div v-else>
        <h2 class="font-extrabold text-lg">
          No products added to the cart yet. Browse products
          <router-link :to="{ name: 'plp' }" class="text-primary"
            >here</router-link
          >
        </h2>
      </div>
    </div>
    <div
      class="basis-full md:basis-1/3 md:border-l border-l-none md:border-l-gray-300 md:pl-4 mt-4 md:mt-0"
      v-if="cart.totalQty"
    >
      <h2 class="font-extrabold text-lg text-start">Detailed Cost Analysis</h2>
      <CartDetails class="mt-8" />
    </div>
  </div>
</template>

<script setup lang="ts">
import CartDetails from "@/components/cart/CartDetails.vue";
import CartItem from "@/components/cart/CartItem.vue";
import { useCartStore } from "@/stores/cart";
import { storeToRefs } from "pinia";

const cartStore = useCartStore();
const { cart } = storeToRefs(cartStore);
</script>
