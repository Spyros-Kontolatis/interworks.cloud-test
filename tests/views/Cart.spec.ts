import { shallowMount } from "@vue/test-utils";
import Cart from "../../src/views/Cart.vue";
import { setActivePinia, createPinia } from "pinia";
import { createTestingPinia } from "@pinia/testing";
import { useCartStore } from "@/stores/cart";
import CartItem from "@/components/cart/CartItem.vue";

beforeEach(() => {
  setActivePinia(createPinia());
});
describe("Cart.vue", () => {
  it("renders the component", () => {
    const wrapper = shallowMount(Cart, {
      global: {
        plugins: [
          createTestingPinia({
            initialState: {
              cart: {
                cart: {
                  id: "1",
                  items: [
                    {
                      product: {
                        id: "1",
                        name: "Product 1",
                        price: 1,
                        description: "A product",
                        image: "some image",
                      },
                      quantity: 2,
                    },
                    {
                      product: {
                        id: "2",
                        name: "Product 2",
                        price: 4,
                        description: "A product",
                        image: "some image",
                      },
                      quantity: 5,
                    },
                  ],
                  totalPrice: 22,
                  totalQty: 7,
                },
              },
            },
            stubActions: false,
          }),
        ],
      },
    });
    useCartStore();
    expect(wrapper.element).toMatchSnapshot();
    const cartItems = wrapper.findAllComponents(CartItem);
    expect(cartItems).toHaveLength(2);
  });
});
