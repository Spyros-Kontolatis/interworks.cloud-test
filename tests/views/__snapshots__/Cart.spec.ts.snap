// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart.vue renders the component 1`] = `
<div
  class="flex flex-col md:flex-row h-full"
>
  <div
    class="basis-full md:basis-2/3 md:overflow-auto"
  >
    <div>
      <h2
        class="font-extrabold text-lg text-start"
      >
        7 Products added to the Cart 
      </h2>
      
      <cart-item-stub
        class="my-8"
        item="[object Object]"
      />
      <cart-item-stub
        class="my-8"
        item="[object Object]"
      />
      
    </div>
  </div>
  <div
    class="basis-full md:basis-1/3 md:border-l border-l-none md:border-l-gray-300 md:pl-4 mt-4 md:mt-0"
  >
    <h2
      class="font-extrabold text-lg text-start"
    >
      Detailed Cost Analysis
    </h2>
    <cart-details-stub
      class="mt-8"
    />
  </div>
</div>
`;
